src/differentiation/compute_jacobian_ad.jl

```diff
@@ -41,48 +41,89 @@ function ForwardColorJacCache(f,x,_chunksize = nothing;
-                                                          undef, num_of_chunks)
+function forwarddiff_color_jacobian(f,
+                x::AbstractArray{<:Number};
+                dx = nothing,
+                colorvec = 1:length(x),
+                sparsity = nothing,
+                jac_prototype = nothing)
+    forwarddiff_color_jacobian(f,x,ForwardColorJacCache(f,x,dx=dx,colorvec=colorvec,sparsity=sparsity),jac_prototype)
+end
 
-    for color_i in 1:maximum(colorvec)
-        for j in 1:length(x)
-            partials[j,color_i] = colorvec[j]==color_i
-        end
-    end
+function forwarddiff_color_jacobian(f,x::AbstractArray{<:Number},jac_cache::ForwardColorJacCache,jac_prototype=nothing)
+    t = jac_cache.t
+    fx = jac_cache.fx
+    dx = jac_cache.dx
+    p = jac_cache.p
+    colorvec = jac_cache.colorvec
+    sparsity = jac_cache.sparsity
+    chunksize = jac_cache.chunksize
+    color_i = 1
+    maxcolor = maximum(colorvec)
 
-    padding_matrix = BitMatrix(undef, length(x), padding_size)
-    partials = hcat(partials, padding_matrix)
+    vecx = vec(x)
+    
+    ncols=length(x)
+    J = jac_prototype isa Nothing ? false .* x .* x' : zero(jac_prototype)
 
-    for i in 1:num_of_chunks
-        partial[:,1] .= partials[:,(i-1)*chunksize+1]
-        for j in 2:chunksize
-            partial[:,j] .= partials[:,(i-1)*chunksize+j]
-        end
-        chunked_partials[i] = Tuple.(eachrow(partial))
+    if !(sparsity isa Nothing)
+        rows_index, cols_index = ArrayInterface.findstructralnz(sparsity)
+        rows_index = [rows_index[i] for i in 1:length(rows_index)]
+        cols_index = [cols_index[i] for i in 1:length(cols_index)]
     end
 
-    chunked_partials
-
+    for i in eachindex(p)
+        partial_i = p[i]
+        t = reshape(Dual{typeof(f)}.(vecx, partial_i),size(t))
+        fx = f(t)
+        if !(sparsity isa Nothing)
+            for j in 1:chunksize
+                dx = vec(partials.(fx, j))
+                pick_inds = [i for i in 1:length(rows_index) if colorvec[cols_index[i]] == color_i]
+                rows_index_c = rows_index[pick_inds]
+                cols_index_c = cols_index[pick_inds]
+                len_rows = length(pick_inds)
+                unused_rows = setdiff(1:ncols,rows_index_c)
+                perm_rows = sortperm([i<=len_rows ? rows_index_c[i] : unused_rows[i-len_rows] for i in 1:ncols])
+                cols_index_c = [i<=len_rows ? cols_index_c[i] : false for i in 1:ncols][perm_rows]
+                Ji = [j==cols_index_c[i] ? dx[i] : false for i in 1:ncols, j in 1:ncols]
+                J = J + Ji
+                color_i += 1
+                (color_i > maxcolor) && return J
+            end
+        else
+            for j in 1:chunksize
+                col_index = (i-1)*chunksize + j
+                (col_index > maxcolor) && return J
+                J = J + partials.(vec(fx), j) .* [(i==col_index ? 1 : 0) for i in 1:ncols]'
```

Comment:
there should be a way to remove the array allocation here.

Reply:
I'm thinking of the delta tensor but that would introduce yet another dependency.

Reply:
The delta tensor? Is that a FillArrays.jl object?

Reply:
I thought there would be some lazy Kronecker delta tensors, but looks like there's no such thing. FillArrays.jl doesn't work, the slice of eye generates sparse vector and its product with a normal array would be sparseCSC which is worse.

Reply:
I can't find any to make a general way of removing this array allocation. If we focus at `StaticArrays`, maybe I can make some overloads for `+` .

Reply:
It would be fine to have a special branch for StaticArrays. How does ForwardDiff do it? It has a special routine for Jacobians that doesn't allocate.